File: src/main/java/server/mecs/proxyservermanager/discord/Discord.java
package server.mecs.proxyservermanager.discord;

import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.JDABuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.events.ReadyEvent;
import net.dv8tion.jda.api.events.guild.member.GuildMemberJoinEvent;
import net.dv8tion.jda.api.events.guild.member.GuildMemberRemoveEvent;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.events.message.priv.PrivateMessageReceivedEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.dv8tion.jda.api.requests.GatewayIntent;
import net.dv8tion.jda.api.utils.ChunkingFilter;
import net.dv8tion.jda.api.utils.MemberCachePolicy;
import net.md_5.bungee.config.Configuration;
import server.mecs.proxyservermanager.ConfigFile;
import server.mecs.proxyservermanager.ProxyServerManager;
import server.mecs.proxyservermanager.commands.discord.McToDiscord;
import server.mecs.proxyservermanager.threads.AccountSync;
import server.mecs.proxyservermanager.threads.AccountUnSync;
import server.mecs.proxyservermanager.threads.CheckSynced;
import server.mecs.proxyservermanager.threads.getIDfromMCID;
import server.mecs.proxyservermanager.utils.getDate;

import javax.security.auth.login.LoginException;
import java.awt.*;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ExecutionException;

public class Discord extends ListenerAdapter {
    public ProxyServerManager plugin = null;

    JDA jda;

    public String token = null;

    public Long guildID = 0L;
    public Guild guild = null;

    public Long receivereportChannelID = 0L;
    public TextChannel receivereportChannel = null;

    public Long staffmessageChannelID = 0L;
    public TextChannel staffmessageChannel = null;

    public EmbedBuilder eb = new EmbedBuilder();

    String date = getDate.getDate();

    public Discord(ProxyServerManager plugin) {
        Configuration config = new ConfigFile(plugin).getConfig();
        token = config.getString("Discord.Token");
        guildID = config.getLong("Discord.Guild");
        receivereportChannelID = config.getLong("Discord.ReportChannel");
        staffmessageChannelID = config.getLong("Discord.StaffChannel");

        this.plugin = plugin;
        setup();
    }

    public void staffmessage(String message) {
        eb.setTitle("**StaffMessage**", null);

        eb.setColor(Color.RED);

        eb.setDescription(date);

        eb.addField("**[Description]**", message, false);

        staffmessageChannel.sendMessage(eb.build()).queue();

        eb.clear();
    }

    public void receivereport(MessageEmbed message) {
        receivereportChannel.sendMessage(message).queue();
    }

    public void shutdown() {
        jda.shutdown();
        plugin.getLogger().info("discord shutdown");
    }

    public void setup() {
        plugin.getLogger().info("discord setup");

        if (token == null) {
            plugin.getLogger().info("Discord token is not initialized.");
            return;
        }

        try {

            jda = JDABuilder
                    .createDefault(token)
                    .addEventListeners(this)
                    .setChunkingFilter(ChunkingFilter.ALL)
                    .setMemberCachePolicy(MemberCachePolicy.ALL)
                    .enableIntents(GatewayIntent.GUILD_MEMBERS)
                    .setActivity(Activity.playing("MECS"))
                    .build();
            jda.awaitReady();

            guild = jda.getGuildById(guildID);
            receivereportChannel = guild.getTextChannelById(receivereportChannelID);
            staffmessageChannel = guild.getTextChannelById(staffmessageChannelID);
        } catch (LoginException | InterruptedException e) {
            e.printStackTrace();
            plugin.getLogger().info(e.getLocalizedMessage());
        }
        plugin.getLogger().info("discord setup done!");
    }

    @Override
    public void onReady(ReadyEvent e) {
        plugin.getLogger().info("discord bot ready");
    }

    @Override
    public void onGuildMemberJoin(GuildMemberJoinEvent e) {
        guild.modifyNickname(e.getMember(), "An_Unlinked_Player").queue();
    }

    @Override
    public void onGuildMemberRemove(GuildMemberRemoveEvent e) {
        try {
            if (CheckSynced.isSynced(plugin, null, e.getUser().getIdLong())) {
                AccountUnSync.AccountUnSync(plugin, null, e.getUser().getIdLong());
            }
        } catch (InterruptedException | ExecutionException ex) {
            ex.printStackTrace();
        }
    }

    @Override
    public void onMessageReceived(MessageReceivedEvent e) {
        if (e.getAuthor().getId().equals(jda.getSelfUser().getId())) {
            return;
        }
        if (e.getChannelType() != ChannelType.TEXT) {
            return;
        }

        String emessage = e.getMessage().getContentDisplay();
        String message = e.getMessage().getContentDisplay().replace("/report", "");
        MessageChannel channel = e.getChannel();

        if (emessage.indexOf("/report") == 0) {
            if (emessage.length() <= 8) {
                eb.setColor(new Color(255, 0, 0));
                eb.setDescription("<@" + e.getAuthor().getId() + ">\n十分な記述がありません。\nThere is not enough description.\n/report <requirement>");
                channel.sendMessage(eb.build()).queue();
                eb.clear();
                return;
            }

            eb.setColor(new Color(0, 255, 0));
            eb.setDescription("<@" + e.getAuthor().getId() + ">レポートを送信しました。\nThe report was sent.");
            channel.sendMessage(eb.build()).queue();
            eb.clear();

            eb.setTitle("**DiscordReport**", null);
            eb.setColor(new Color(0, 255, 0));
            eb.setDescription(date);
            eb.addField("**[Description]**", "**[Sender]** <@" + e.getAuthor().getId() + ">\n \n`" + message + "`", false);
            receivereport(eb.build());
            eb.clear();
        }
    }

    @Override
    public void onPrivateMessageReceived(PrivateMessageReceivedEvent e) {
        if (e.getMessage().getAuthor() == jda.getSelfUser()) return;
        try {
            Integer.parseInt(e.getMessage().getContentDisplay());
        } catch (NumberFormatException ex) {
            e.getMessage().getPrivateChannel().sendMessage("Failed to account sync.\nThe ID is not the correct number.").queue();
            return;
        }

        String player = getKeyByValue(McToDiscord.number, Integer.parseInt(e.getMessage().getContentDisplay()));
        Long id = e.getMessage().getAuthor().getIdLong();
        Integer ID = Integer.parseInt(e.getMessage().getContentDisplay());

        if (!(McToDiscord.number.containsValue(ID))) {
            e.getMessage().getPrivateChannel().sendMessage("Failed to account sync.\nPlease try again.").queue();
            return;
        }

        try {
            if (CheckSynced.isSynced(plugin, player, null)) {
                e.getMessage().getPrivateChannel().sendMessage("Failed to account sync.\nApparently your account has already have sync.").queue();
                return;
            }
        } catch (NullPointerException | InterruptedException | ExecutionException ex) {
            e.getMessage().getPrivateChannel().sendMessage("Failed to account sync.\nPlease try again.").queue();
            return;
        }

        try {
            if (CheckSynced.isSynced(plugin, null, e.getMessage().getAuthor().getIdLong())) {
                e.getMessage().getPrivateChannel().sendMessage("Failed to account sync.\nApparently your account has already have sync.").queue();
                return;
            }
        } catch (InterruptedException | ExecutionException ex) {
            ex.printStackTrace();
        }

        try {
            AccountSync.AccountSync(plugin, player, id);
            guild.modifyNickname(guild.getMemberById(id), player).queue();
            guild.addRoleToMember(id, guild.getRoleById(753582521685377034L)).queue();
            McToDiscord.number.remove(player);
            e.getMessage().getPrivateChannel().sendMessage("Successfully synced with your Minecraft account.").queue();
        } catch (Exception ex) {
            ex.printStackTrace();
            e.getMessage().getPrivateChannel().sendMessage("Failed to account sync.\nPlease report to the Staff Team.").queue();
        }
    }

    public static <T, E> T getKeyByValue(Map<T, E> map, E value) {
        for (Map.Entry<T, E> entry : map.entrySet()) {
            if (Objects.equals(value, entry.getValue())) {
                return entry.getKey();
            }
        }
        return null;
    }

    public void removeRole(String mcid) throws InterruptedException {
        guild.removeRoleFromMember(getIDfromMCID.getIDfromMCID(plugin, mcid), guild.getRoleById(753582521685377034L)).queue();
    }
}

File: src/main/java/server/mecs/proxyservermanager/discord/Discord.kt
package server.mecs.proxyservermanager.discord

import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.JDABuilder
import net.dv8tion.jda.api.entities.ChannelType
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.events.ReadyEvent
import net.dv8tion.jda.api.events.guild.member.GuildMemberJoinEvent
import net.dv8tion.jda.api.events.guild.member.GuildMemberRemoveEvent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.events.message.priv.PrivateMessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.utils.ChunkingFilter
import net.dv8tion.jda.api.utils.MemberCachePolicy
import server.mecs.proxyservermanager.ProxyServerManager
import server.mecs.proxyservermanager.ProxyServerManager.Companion.es
import server.mecs.proxyservermanager.ProxyServerManager.Companion.getDate
import server.mecs.proxyservermanager.commands.discord.McToDiscord
import server.mecs.proxyservermanager.database.DiscordData.isSynced
import server.mecs.proxyservermanager.database.DiscordData.syncAccount
import server.mecs.proxyservermanager.database.DiscordData.unsyncAccount
import server.mecs.proxyservermanager.database.MySQLManager
import server.mecs.proxyservermanager.database.PlayerData.getIDfromMCID
import java.awt.Color
import java.util.concurrent.ExecutionException
import javax.security.auth.login.LoginException

class Discord(private val plugin: ProxyServerManager) : ListenerAdapter() {
    var jda: JDA? = null
    var token: String? = null
    var guildID = 0L
    var guild: Guild? = null
    var receivereportChannelID = 0L
    var receivereportChannel: TextChannel? = null
    var staffmessageChannelID = 0L
    var staffmessageChannel: TextChannel? = null
    var eb = EmbedBuilder()
    fun staffmessage(message: String?) {
        eb.setTitle("**StaffMessage**", null)
        eb.setColor(Color.RED)
        eb.setDescription(getDate().toString())
        eb.addField("**[Description]**", message, false)
        staffmessageChannel!!.sendMessage(eb.build()).queue()
        eb.clear()
    }

    fun receivereport(message: MessageEmbed?) {
        receivereportChannel!!.sendMessage(message!!).queue()
    }

    fun shutdown() {
        jda!!.shutdown()
        plugin.logger.info("discord shutdown")
    }

    fun setup() {
        plugin.logger.info("discord setup")
        if (token == null) {
            plugin.logger.info("Discord token is not initialized.")
            return
        }
        try {
            jda = JDABuilder
                    .createDefault(token)
                    .addEventListeners(this)
                    .setChunkingFilter(ChunkingFilter.ALL)
                    .setMemberCachePolicy(MemberCachePolicy.ALL)
                    .enableIntents(GatewayIntent.GUILD_MEMBERS)
                    .build()
            jda!!.awaitReady()
            guild = jda!!.getGuildById(guildID)
            receivereportChannel = guild!!.getTextChannelById(receivereportChannelID)
            staffmessageChannel = guild!!.getTextChannelById(staffmessageChannelID)
        } catch (e: LoginException) {
            e.printStackTrace()
            plugin.logger.info(e.localizedMessage)
        } catch (e: InterruptedException) {
            e.printStackTrace()
            plugin.logger.info(e.localizedMessage)
        }
        plugin.logger.info("discord setup done!")
    }

    override fun onReady(e: ReadyEvent) {
        plugin.logger.info("discord bot ready")
    }

    override fun onGuildMemberJoin(e: GuildMemberJoinEvent) {
        es.execute { guild!!.modifyNickname(e.member, "An_Unlinked_Player").queue() }
    }

    override fun onGuildMemberRemove(e: GuildMemberRemoveEvent) {
        es.execute {
            MySQLManager(plugin, "Discord GuildMemberRemove").use { mysql ->
                val id = e.user.idLong
                try {
                    if (isSynced(mysql, null, id)) {
                        unsyncAccount(mysql, null, id)
                    }
                } catch (ex: InterruptedException) {
                    ex.printStackTrace()
                } catch (ex: ExecutionException) {
                    ex.printStackTrace()
                }
            }
        }
    }

    override fun onMessageReceived(e: MessageReceivedEvent) {
        es.execute {
            if (e.author.id == jda!!.selfUser.id) {
                return@execute
            }
            if (e.channelType != ChannelType.TEXT) {
                return@execute
            }
            val emessage = e.message.contentDisplay
            val message = e.message.contentDisplay.replace("/report", "")
            val channel = e.channel
            val mention = "<@${e.author.id}>"
            if (emessage.indexOf("/report") == 0) {
                if (emessage.length <= 8) {
                    eb.setColor(Color(255, 0, 0))
                    eb.setDescription("""
    $mention
    十分な記述がありません。
    There is not enough description.
    /report <requirement>
    """.trimIndent())
                    channel.sendMessage(eb.build()).queue()
                    eb.clear()
                    return@execute
                }
                eb.setColor(Color(0, 255, 0))
                eb.setDescription("""
    ${mention}レポートを送信しました。。
    Your report was sent.
    """.trimIndent())
                channel.sendMessage(eb.build()).queue()
                eb.clear()
                eb.setTitle("**DiscordReport**", null)
                eb.setColor(Color(0, 255, 0))
                eb.setDescription(getDate().toString())
                eb.addField("**[Description]**", """**[Sender]** $mention

`$message`""", false)
                receivereport(eb.build())
                eb.clear()
            }
        }
    }

    override fun onPrivateMessageReceived(e: PrivateMessageReceivedEvent) {
        es.execute {
            if (e.message.author === jda!!.selfUser) return@execute
            try {
                e.message.contentDisplay.toInt()
            } catch (ex: NumberFormatException) {
                e.message.privateChannel.sendMessage("Failed to account sync.\nThe ID is not the correct number.").queue()
                return@execute
            }
            val player = getKeyByValue(McToDiscord.number, e.message.contentDisplay.toInt())!!
            val id = e.message.author.idLong
            val ID = e.message.contentDisplay.toInt()
            if (!McToDiscord.number.containsValue(ID)) {
                e.message.privateChannel.sendMessage("Failed to account sync.\nPlease try again.").queue()
                return@execute
            }
            MySQLManager(plugin, "Discord PrivateMessageReceived").use { mysql ->
                try {
                    if (isSynced(mysql, player, null)) {
                        e.message.privateChannel.sendMessage("Failed to account sync.\nApparently your account has already have sync.").queue()
                        return@execute
                    }
                } catch (ex: NullPointerException) {
                    e.message.privateChannel.sendMessage("Failed to account sync.\nPlease try again.").queue()
                    return@execute
                } catch (ex: InterruptedException) {
                    e.message.privateChannel.sendMessage("Failed to account sync.\nPlease try again.").queue()
                    return@execute
                } catch (ex: ExecutionException) {
                    e.message.privateChannel.sendMessage("Failed to account sync.\nPlease try again.").queue()
                    return@execute
                }
                try {
                    if (isSynced(mysql, null, e.message.author.idLong)) {
                        e.message.privateChannel.sendMessage("Failed to account sync.\nApparently your account has already have sync.").queue()
                        return@execute
                    }
                } catch (ex: InterruptedException) {
                    ex.printStackTrace()
                } catch (ex: ExecutionException) {
                    ex.printStackTrace()
                }
                try {
                    syncAccount(mysql, player, id)
                    if (isSynced(mysql, null, id)) {
                        guild!!.modifyNickname(guild!!.getMemberById(id)!!, player).queue()
                        guild!!.addRoleToMember(id, guild!!.getRoleById(753582521685377034L)!!).queue()
                        McToDiscord.number.remove(player)
                        e.message.privateChannel.sendMessage("Successfully synced with your Minecraft account.").queue()
                    }
                    e.message.privateChannel.sendMessage("Failed to account sync.\nPlease report to the Staff Team.").queue()
                } catch (ex: Exception) {
                    ex.printStackTrace()
                    e.message.privateChannel.sendMessage("Failed to account sync.\nPlease report to the Staff Team.").queue()
                }
            }
        }
    }

    fun removeRole(mcid: String?) {
        es.execute {
            MySQLManager(plugin, "Discord removeRole").use { mysql ->
                guild?.removeRoleFromMember(getIDfromMCID(mysql, mcid!!)!!,
                        guild!!.getRoleById(753582521685377034L)!!)?.queue()
            }
        }
    }

    fun <T, E> getKeyByValue(map: Map<T, E>, value: E): T? {
        for ((key, value1) in map) {
            if (value == value1) {
                return key
            }
        }
        return null
    }
}